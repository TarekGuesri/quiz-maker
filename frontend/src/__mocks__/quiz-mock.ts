import { Quiz } from "src/types";

export const quizMock: Quiz = {
  title: "Mock Quiz",
  description: "This a mock quiz for testing",
  code: "sv5F7ohiX",
  questions: [
    {
      content: "Question 1",
      answers: [
        {
          text: "1",
          id: "617d505176471ee1e39fd324",
        },
        {
          text: "2",
          id: "617d505176471ee1e39fd325",
        },
        {
          text: "3",
          id: "617d505176471ee1e39fd326",
        },
        {
          text: "4",
          id: "617d505176471ee1e39fd327",
        },
      ],
      id: "617d505176471ee1e39fd374",
    },
    {
      content: "Question 2",
      answers: [
        {
          text: "1",
          id: "617d505176471ee1e39fd32c",
        },
        {
          text: "2",
          id: "617d505176471ee1e39fd32d",
        },
        {
          text: "3",
          id: "617d505176471ee1e39fd32e",
        },
        {
          text: "4",
          id: "617d505176471ee1e39fd32f",
        },
      ],
      id: "617d505176471ee1e39fd375",
    },
    {
      content: "Question 3",
      answers: [
        {
          text: "1",
          id: "617d505176471ee1e39fd334",
        },
        {
          text: "2",
          id: "617d505176471ee1e39fd335",
        },
        {
          text: "3",
          id: "617d505176471ee1e39fd336",
        },
        {
          text: "4",
          id: "617d505176471ee1e39fd337",
        },
      ],
      id: "617d505176471ee1e39fd376",
    },
    {
      content: "Question 4",
      answers: [
        {
          text: "1",
          id: "617d505176471ee1e39fd33c",
        },
        {
          text: "2",
          id: "617d505176471ee1e39fd33d",
        },
        {
          text: "3",
          id: "617d505176471ee1e39fd33e",
        },
        {
          text: "4",
          id: "617d505176471ee1e39fd33f",
        },
      ],
      id: "617d505176471ee1e39fd377",
    },
    {
      content: "Question 5",
      answers: [
        {
          text: "1",
          id: "617d505176471ee1e39fd344",
        },
        {
          text: "2",
          id: "617d505176471ee1e39fd345",
        },
        {
          text: "3",
          id: "617d505176471ee1e39fd346",
        },
        {
          text: "4",
          id: "617d505176471ee1e39fd347",
        },
      ],
      id: "617d505176471ee1e39fd378",
    },
    {
      content: "Question 6",
      answers: [
        {
          text: "1",
          id: "617d505176471ee1e39fd34c",
        },
        {
          text: "2",
          id: "617d505176471ee1e39fd34d",
        },
        {
          text: "3",
          id: "617d505176471ee1e39fd34e",
        },
        {
          text: "4",
          id: "617d505176471ee1e39fd34f",
        },
      ],
      id: "617d505176471ee1e39fd379",
    },
    {
      content: "Question 7",
      answers: [
        {
          text: "1",
          id: "617d505176471ee1e39fd354",
        },
        {
          text: "2",
          id: "617d505176471ee1e39fd355",
        },
        {
          text: "3",
          id: "617d505176471ee1e39fd356",
        },
        {
          text: "4",
          id: "617d505176471ee1e39fd357",
        },
      ],
      id: "617d505176471ee1e39fd37a",
    },
    {
      content: "Question 8",
      answers: [
        {
          text: "1",
          id: "617d505176471ee1e39fd35c",
        },
        {
          text: "2",
          id: "617d505176471ee1e39fd35d",
        },
        {
          text: "3",
          id: "617d505176471ee1e39fd35e",
        },
        {
          text: "4",
          id: "617d505176471ee1e39fd35f",
        },
      ],
      id: "617d505176471ee1e39fd37b",
    },
    {
      content: "Question 9",
      answers: [
        {
          text: "1",
          id: "617d505176471ee1e39fd364",
        },
        {
          text: "2",
          id: "617d505176471ee1e39fd365",
        },
        {
          text: "3",
          id: "617d505176471ee1e39fd366",
        },
        {
          text: "4",
          id: "617d505176471ee1e39fd367",
        },
      ],
      id: "617d505176471ee1e39fd37c",
    },
    {
      content: "Question 10",
      answers: [
        {
          text: "1",
          id: "617d505176471ee1e39fd36c",
        },
        {
          text: "2",
          id: "617d505176471ee1e39fd36d",
        },
        {
          text: "3",
          id: "617d505176471ee1e39fd36e",
        },
        {
          text: "4",
          id: "617d505176471ee1e39fd36f",
        },
      ],
      id: "617d505176471ee1e39fd37d",
    },
  ],
  id: "617d505176471ee1e39fd388",
};
