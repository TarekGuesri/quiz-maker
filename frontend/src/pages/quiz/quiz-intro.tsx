import React, { useState } from "react";
import { makeStyles } from "@mui/styles";
import { Theme } from "@mui/material";
import ligthBulbImage from "src/assets/images/quiz-light-bulb.png";

const useStyles = makeStyles((theme: Theme) => ({
  ligthBulbImage: {
    width: "75px",
    [theme.breakpoints.down("md")]: {
      width: "50px",
    },
  },
  ligthBulbBackground: {
    border: `9px solid ${theme.palette.secondary.main}`,
    backgroundColor: `${theme.palette.primary.main}`,
    display: "inline-flex",
    padding: "18px",
    borderRadius: "100px",
  },
}));

export const QuizIntro: React.FC = () => {
  const classes = useStyles();
  // We use this to hide div style (border and backround) when image is loading
  const [imageLoading, setImageLoading] = useState(true);

  return (
    <>
      <div className={imageLoading ? "" : classes.ligthBulbBackground}>
        <img
          src={ligthBulbImage}
          alt=""
          className={classes.ligthBulbImage}
          onLoad={() => setImageLoading(false)}
        />
      </div>
    </>
  );
};
