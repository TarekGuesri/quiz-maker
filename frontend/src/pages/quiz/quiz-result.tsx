import React from "react";
import Typography from "@mui/material/Typography";
import Box from "@mui/material/Box";
import { CircularProgressWithLabel } from "src/components/progress/circular-progress-with-label";

interface QuizResultProps {
  score: number;
}

export const QuizResult: React.FC<QuizResultProps> = ({ score }) => {
  return (
    <>
      <Typography color="secondary" variant="h4" component="div" mb={4}>
        Quiz Result
      </Typography>

      <Box mb={4}>
        <CircularProgressWithLabel value={score} />
      </Box>

      <Typography variant="h6" mb={1}>
        You have completed the quiz.
      </Typography>
      <Typography variant="h6">
        You answered <strong> {score}%</strong> of the questions correctly.
      </Typography>
    </>
  );
};
