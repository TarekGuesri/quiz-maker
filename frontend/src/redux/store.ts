import { Action, ThunkAction, configureStore } from "@reduxjs/toolkit";

import ui from "./ui/ui-slice";

export const store = configureStore({
  reducer: {
    ui,
  },
  devTools: true,
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
