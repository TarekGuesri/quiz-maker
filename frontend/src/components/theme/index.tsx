import React, { FC } from 'react';
import CssBaseline from '@mui/material/CssBaseline';
import PropTypes from 'prop-types';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { useAppSelector } from 'src/redux/hooks';
import varelaRound from 'src/assets/fonts/VarelaRound-Regular.ttf';

export const Theme: FC = (props) => {
  const { darkMode } = useAppSelector((state) => state.ui);

  const theme = createTheme({
    palette: {
      mode: darkMode ? 'dark' : 'light',
      background: {
        default: darkMode ? '#333' : '#f7f7f7',
        paper: darkMode ? '#333' : '#fbfbfb',
      },
      primary: {
        main: '#8e49de',
        // contrastText:
      },
      secondary: {
        main: '#f9de4a',
      },
      contrastThreshold: 3,
      tonalOffset: 0.2,
    },
    typography: {
      fontFamily: 'Varela Round',
    },
    components: {
      MuiCssBaseline: {
        styleOverrides: `
          @font-face {
            font-family: 'Varela Round';
            font-style: normal;
            font-display: swap;
            font-weight: 400;
            src: url(${varelaRound}) format('truetype');
        `,
      },
    },
  });

  return (
    <ThemeProvider theme={theme}>
      {props.children}
      <CssBaseline />
    </ThemeProvider>
  );
};

Theme.propTypes = {
  children: PropTypes.node.isRequired,
};
