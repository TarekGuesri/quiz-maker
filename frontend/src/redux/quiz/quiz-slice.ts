import axios, { AxiosError, AxiosResponse } from "axios";
import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AppThunk, RootState } from "src/redux/store";
import { QuizState, Quiz } from "src/types";

const initialState: QuizState = {
  quiz: {} as Quiz,
  questionIndex: 0,
  selectedAnswers: [],
  isLoading: true,
  quizStarted: false,
  quizTimer: "00:00",
  isSubmitting: false,
  quizResult: -1, // After getting the result from backend, it becomes 0 or higher
  errorMessage: "",
};

export const quizSlice = createSlice({
  name: "quiz",
  initialState,
  reducers: {
    resetState: () => initialState,
    setQuiz: (state, action: PayloadAction<Quiz>) => {
      state.quiz = { ...action.payload };
      state.isLoading = false;
    },
    getQuizFail: (state, action: PayloadAction<string>) => {
      state.errorMessage = action.payload;
      state.isLoading = false;
    },
    startQuiz: (state) => {
      state.quizStarted = true;
    },
    changeTimer: (state, action: PayloadAction<string>) => {
      state.quizTimer = action.payload;
    },
    nextQuestion: (state) => {
      if (state.questionIndex + 1 < state.quiz.questions.length) {
        state.questionIndex++;
      }
    },
    selectAnswer: (state, action: PayloadAction<string>) => {
      state.selectedAnswers[state.questionIndex] = action.payload;
    },
  },
});

export const {
  changeTimer,
  setQuiz,
  getQuizFail,
  resetState,
  startQuiz,
  nextQuestion,
  selectAnswer,
} = quizSlice.actions;

export const selectQuiz = (state: RootState) => state.quiz;

export const getQuizByCode =
  (quizCode: string): AppThunk =>
  async (dispatch) => {
    try {
      const res: AxiosResponse = await axios.get(`quizzes/${quizCode}`);

      dispatch(setQuiz(res.data));
    } catch (error) {
      const { response } = error as AxiosError;

      const errorMessage = response?.data || "Something unexpected happend!";

      dispatch(getQuizFail(errorMessage));
    }
  };

export const getQuizResult = (): AppThunk => async (dispatch, getState) => {
  const { quiz } = selectQuiz(getState());
  try {
    const res: AxiosResponse = await axios.post(`quizzes/result/${quiz.code}`);

    console.log(res.data);
  } catch (error) {
    const { response } = error as AxiosError;

    const errorMessage = response?.data || "Something unexpected happend!";

    dispatch(getQuizFail(errorMessage));
  }
};

export default quizSlice.reducer;
