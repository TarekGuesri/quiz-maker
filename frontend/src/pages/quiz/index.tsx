import React, { useEffect } from "react";
import { useParams } from "react-router-dom";
import { useAppSelector } from "src/redux/hooks";
import Box from "@mui/material/Box";
import Fade from "@mui/material/Fade";
import Typography from "@mui/material/Typography";
import { CardBox } from "src/components/cards/card-box";
import { store } from "src/redux/store";
import { getQuizByCode } from "src/redux/quiz/quiz-slice";
import { Loading } from "src/components/layout/loading";
import { QuizIntro } from "./quiz-intro";

const Quiz: React.FC = () => {
  const { quiz, isLoading, quizStarted, errorMessage } = useAppSelector(
    (state) => state.quiz,
  );

  const { quizCode } = useParams<{ quizCode: string }>();

  // Getting quiz
  useEffect(() => {
    store.dispatch(getQuizByCode(quizCode));
  }, []);

  if (isLoading) {
    return <Loading />;
  }

  if (errorMessage) {
    return (
      <CardBox>
        <Typography color="secondary" variant="h5" component="div" mb={3}>
          {errorMessage}
        </Typography>
        {errorMessage === "Quiz not found!" && (
          <Typography component="div">
            Wrong link or quiz has been deleted!
          </Typography>
        )}
      </CardBox>
    );
  }

  return (
    <CardBox>
      <Typography color="secondary" variant="h4" component="div" mb={1}>
        {quiz?.title}
      </Typography>
      <Box mt={3}>
        {!quizStarted ? (
          <Fade
            in={!quizStarted}
            style={{ transitionDelay: "100ms" }}
            mountOnEnter
            unmountOnExit
          >
            <div>
              <QuizIntro />
            </div>
          </Fade>
        ) : (
          <></>
        )}

        <Fade
          in={quizStarted}
          style={{ transitionDelay: "1000ms" }}
          mountOnEnter
          unmountOnExit
        >
          <div>
            <>QuizStarted</>
          </div>
        </Fade>
      </Box>
    </CardBox>
  );
};

export default Quiz;
