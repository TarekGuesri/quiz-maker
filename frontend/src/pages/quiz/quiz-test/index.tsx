import React, { useEffect } from "react";

import { makeStyles } from "@mui/styles";
import Typography from "@mui/material/Typography";
import Button from "@mui/material/Button";
import { Theme } from "@mui/material";
import { store } from "src/redux/store";
import { startQuiz, resetState } from "src/redux/quiz/quiz-slice";
import { Timer } from "./timer";

const useStyles = makeStyles((theme: Theme) => ({
  button: {
    width: "150px",
    padding: "10px",
  },
}));

export const QuizTest: React.FC = () => {
  const classes = useStyles();

  useEffect(() => {
    return () => {
      store.dispatch(resetState());
    };
  }, []);

  return (
    <>
      <Timer />

      <Typography mb={4}>
        To get started with the quiz, click on the button below
      </Typography>

      <Button
        variant="contained"
        color="primary"
        className={classes.button}
        onClick={() => store.dispatch(startQuiz())}
        // disabled={questions.length < 2}
      >
        Start Quiz
      </Button>
    </>
  );
};
