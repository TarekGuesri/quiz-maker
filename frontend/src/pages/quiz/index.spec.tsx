import React from "react";
import axios from "axios";
import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { store } from "src/redux/store";
import { Router } from "react-router-dom";
import { history } from "src/utils/history";
import { Provider } from "react-redux";
import { Theme } from "src/components/theme";
import { setQuiz, resetState } from "src/redux/quiz/quiz-slice";
import { quizMock } from "src/__mocks__/quiz-mock";
import Quiz from ".";

jest.mock("axios");

describe("src/pages/quiz/index.tsx", () => {
  afterEach(() => cleanup());
  beforeEach(() => {
    store.dispatch(resetState());
  });

  test("displaying 'Wrong link or quiz has been deleted!' when error message is 'Quiz not found!'", async () => {
    const errorMessage = "Quiz not found!";

    // Checking failed request
    (axios as jest.Mocked<typeof axios>).get.mockRejectedValue({
      response: { data: errorMessage },
    });

    const renderedComponent = render(
      <Provider store={store}>
        <Router history={history}>
          <Theme>
            <Quiz />
          </Theme>
        </Router>
      </Provider>,
    );

    await waitFor(() =>
      expect(screen.queryByText(errorMessage)).not.toBeNull(),
    );

    expect(
      screen.queryByText(/wrong link or quiz has been deleted!/i),
    ).not.toBeNull(),
      expect(renderedComponent.container).toMatchSnapshot();
  });
});
