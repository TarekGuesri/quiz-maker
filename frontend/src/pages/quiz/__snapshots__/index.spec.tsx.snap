// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/quiz/index.tsx displaying 'Wrong link or quiz has been deleted!' when error message is 'Quiz not found!' 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root makeStyles-root-3 css-4nwtfp-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root makeStyles-cardContent-4 css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiTypography-root MuiTypography-h5 css-c5bhyc-MuiTypography-root"
      >
        Quiz not found!
      </div>
      <div
        class="MuiTypography-root MuiTypography-body1 css-jqqqr0-MuiTypography-root"
      >
        Wrong link or quiz has been deleted!
      </div>
    </div>
  </div>
</div>
`;
